src/runtime/cuda/cuda_device_api.cc

```diff
@@ -214,10 +234,33 @@ class CUDADeviceAPI final : public DeviceAPI {
   bool PinData(void* ptr, size_t nbytes) override {
     // prevent users from pinning empty tensors or graphs
     if (ptr == nullptr || nbytes == 0) return false;
+    TensorDispatcher* tensor_dispatcher = TensorDispatcher::Global();
+    // Minimize the pinned memory pool allocated by backend (via tensoradapter)
+    // to preserve enough memory for DGL inherited in-place pin-memory operation
+    if (tensor_dispatcher->IsAvailable()) {
+      tensor_dispatcher->CUDAHostAllocEmptyCache();
+    }
     CUDA_CALL(cudaHostRegister(ptr, nbytes, cudaHostRegisterDefault));
     return true;
   }
 
+  void* AllocPinnedDataSpace(
+      size_t nbytes, void** ctx, void** deleter) override {
+    // prevent pinning empty tensors or graphs
+    if (nbytes == 0) return nullptr;
+    TensorDispatcher* tensor_dispatcher = TensorDispatcher::Global();
+    CHECK(tensor_dispatcher->IsAvailable())
+        << "CachingHost allocator only available with TensorAdapter.";
```

Comment:
So `graph.pin_memory()` will only work if tensor adapter is available?  I think a better error message is needed if this is the case, like telling the users to update the PyTorch version to what we support.  @frozenbugs what do you think?

Reply:
>So graph.pin_memory() will only work if tensor adapter is available? 

Yes.

I updated the msg to `CachingHostAllocator is not available in the current backend PyTorch. Please update the PyTorch version to 1.11+`.
We can circulate from here.